contracts/token/ERC721/ERC721BasicToken.sol

```diff
@@ -0,0 +1,284 @@
+  * @param _from current owner of the token
+  * @param _to address to receive the ownership of the given token ID
+  * @param _tokenId uint256 ID of the token to be transferred
+  */
+  function transferFrom(address _from, address _to, uint256 _tokenId) public canTransfer(_tokenId) {
+    clearApprovalAndTransfer(_from, _to, _tokenId, "", false);
+  }
+
+  /**
+  * @dev Safely transfers the ownership of a given token ID to another address
+  * @dev If the target address is a contract, it must implement `onERC721Received`,
+  *  which is called upon a safe transfer, and return the magic value
+  *  `bytes4(keccak256("onERC721Received(address,uint256,bytes)"))`; otherwise,
+  *  the transfer is reverted.
+  * @dev Requires the msg sender to be the owner, approved, or operator
+  * @param _from current owner of the token
+  * @param _to address to receive the ownership of the given token ID
+  * @param _tokenId uint256 ID of the token to be transferred
+  */
+  function safeTransferFrom(address _from, address _to, uint256 _tokenId) public canTransfer(_tokenId) {
+    clearApprovalAndTransfer(_from, _to, _tokenId, "", true);
+  }
+
+  /**
+  * @dev Safely transfers the ownership of a given token ID to another address
+  * @dev If the target address is a contract, it must implement `onERC721Received`,
+  *  which is called upon a safe transfer, and return the magic value
+  *  `bytes4(keccak256("onERC721Received(address,uint256,bytes)"))`; otherwise,
+  *  the transfer is reverted.
+  * @dev Requires the msg sender to be the owner, approved, or operator
+  * @param _from current owner of the token
+  * @param _to address to receive the ownership of the given token ID
+  * @param _tokenId uint256 ID of the token to be transferred
+  * @param _data bytes data to send along with a safe transfer check
+  */
+  function safeTransferFrom(address _from, address _to, uint256 _tokenId, bytes _data) public canTransfer(_tokenId) {
+    clearApprovalAndTransfer(_from, _to, _tokenId, _data, true);
+  }
+
+  /**
+   * @dev Returns whether the given spender can transfer a given token ID
+   * @param _spender address of the spender to query
+   * @param _tokenId uint256 ID of the token to be transferred
+   * @return bool whether the msg.sender is approved for the given token ID,
+   *  is an operator of the owner, or is the owner of the token
+   */
+  function isApprovedOrOwner(address _spender, uint256 _tokenId) internal view returns (bool) {
+    address owner = ownerOf(_tokenId);
+    return _spender == owner || getApproved(_tokenId) == _spender || isApprovedForAll(owner, _spender);
+  }
+
+  /**
+  * @dev Internal function to mint a new token
+  * @dev Reverts if the given token ID already exists
+  * @param _to The address that will own the minted token
+  * @param _tokenId uint256 ID of the token to be minted by the msg.sender
+  */
+  function doMint(address _to, uint256 _tokenId) internal {
```

Comment:
Consider adding the modifier `onlyOwner` (implying that the contract should be `Ownable`) so that not anyone can mint tokens.

Reply:
That'll probably come in a Mintable extension (similar to how ERC20 is structured). I didn't want to enforce ownership for the base implementation.

Reply:
any reason we went with `doMint` rather than `_mint`? likewise for `doBurn` vs `_burn`. and likewise for the various `do*` internal functions throughout the contracts.

Reply:
Personal preference exclusively. As we were using leading underscores for parameters in functions, I wanted to use something different for the internal methods that actually executed the action, and I had seen the *do* pattern being used in other languages. But I can switch to underscore if preferred.

Reply:
You see that `do` thing come up occasionally. Personally I found it adds little semantically. Every method "does" something, that's why they're methods, start using it often and you can find whole objects full of "doMethods".

Underscore is fairly common in solidity as some kind of "private" be it as a prefix or post depending on the needs. I was able to understand this usage quickly on my first exposure to https://github.com/OpenZeppelin/zeppelin-solidity/blob/1eea95f9acebada0360f1f4be7c442325db27fa6/contracts/token/ERC721/ERC721Token.sol#L120

Initial response was "Why do they have a mint method which isn't actually used and this isn't a Mintable?" Then I see `addToken` and that I'm being given an opportunity to inject other logic into this minting flow (something I functionally require to add extras to an NFT at mint-time).

Reply:
personally I prefer the `_` approach since it's separate, in my head, from the function arguments, but I don't have a strong opinion. My only note is that this will very much inform future contract standard—I'd expect future additions to have `do*` instead of `_*` prefixes as people read the example code and work off of it.